### ISSUE TITLE
Render Thread hangs when server (running in background) GCs

### ISSUE THREAD

**Issue by PitchBlackNights** [RaphiMC/ImmediatelyFast#100]
## Issue Description
When trying to enable a shader pack or configure one, the entire game freezes, then the screen renderer hangs.

## Why
I have no idea what is causing this, and it seems to happen randomly. There is no log output or any other sign. This could even affect other GUI elements, but I've only seen it for Iris.

## Versions
Minecraft 1.19.4
Iris 1.6.5
ImmediatelyFast 1.1.26

**Comment by RaphiMC** [comment]
Which other mods do you have installed? Also please send you PC specs

**Comment by PitchBlackNights** [comment]
@RaphiMC 
I am currently away from my computer, but I will send those items as soon as I can. Should be able to send them around 8pm CST.

**Comment by PitchBlackNights** [comment]
@RaphiMC 
## Laptop specs
```
Processor: 11th Gen Intel(R) Core(TM) i5-11400H @ 2.70GHz
RAM: 16 GB
GPU: NVIDIA GeForce RTX 3050 Ti Laptop GPU
GPU Driver: NVIDIA Game Ready - Version: 536.67
```

## Mods installed

<details>
<summary>Mods</summary>

	- ambientsounds 5.2.19
	- animatica 0.6+1.20
	- appleskin 2.4.4+mc1.19.4
	- badpackets 0.4.1
	- betterf3 6.0.2
	- boatiview 0.0.4
	- c2me 0.2.0+alpha.10.65
	- carpet 1.4.101+v230319
	- chat_heads 0.10.19
	- citresewn 1.1.3+1.19.4
	- cloth-config 10.1.105
	- continuity 3.0.0-beta.3+1.19.4
	- craftpresence 2.2.1
	- creativecore 2.10.12
	- debugify 1.19.4+2.2
	- dynamicfps 2.2.0
	- entity_model_features 0.2.13
	- entity_texture_features 4.4.4
	- entityculling 1.6.2-mc1.19.4
	- exordium 1.1.2-mc1.19.4
	- fabric-api 0.86.1+1.19.4
	- fabric-language-kotlin 1.10.8+kotlin.1.9.0
	- fabricloader 0.14.22
	- fastload 3.4.0
	- ferritecore 5.2.0
	- immediatelyfast 1.1.26+1.19.4
	- indium 1.0.19+mc1.19.4
	- iris 1.6.5
	- jade 10.4.0
	- java 17
	- krypton 0.2.2
	- lambdynlights 2.3.1+1.19.4
	- lazydfu 0.1.3
	- lithium 0.11.1
	- memoryleakfix 1.1.1
	- minecraft 1.19.4
	- modelfix 1.10
	- modmenu 6.3.1
	- moreculling 1.19.4-0.17.0
	- nochatreports 1.19.4-v2.1.6
	- optigui 2.1.5
	- presencefootsteps 1.8.2
	- reeses-sodium-options 1.6.3+mc1.19.4-build.90
	- smoothboot 1.19.4-1.7.0
	- sodium 0.4.10+build.24
	- sodium-extra 0.4.18+mc1.19.4-build.100
	- sound_physics_remastered 1.19.4-1.1.1
	- spark 1.10.37
	- starlight 1.1.1+fabric.ae22326
	- voicechat 1.19.4-2.4.16
	- yet_another_config_lib_v3 3.1.0+1.19.4
	- zoomify 2.9.4

</details>

**Comment by hexadecimal233** [comment]
I think update your mods to 1.20.1 could solve this as it didnt happen to me. Lots of the mods above do not support 1.19.4 anymore

**Comment by PitchBlackNights** [comment]
I've done some testing, and it seems that the game freezes when a Minecraft server collects garbage. Don't ask why I was running a Minecraft server in the background.
The game does not freeze when IF is disabled and the server collects garbage.

**Comment by PitchBlackNights** [comment]
> I think update your mods to 1.20.1 could solve this as it didnt happen to me. Lots of the mods above do not support 1.19.4 anymore

I think some of my mods aren't updated to 1.20.1 yet. Though, I have not confirmed this. I just chose 1.19.4 cause I know that all the mods I wanted, had 1.19.4 versions.

**Comment by RaphiMC** [comment]
Can you upload a game log

**Comment by Raconteur32** [comment]
Also, can you tell us the launcher you use please ?

**Comment by PitchBlackNights** [comment]
Sorry for the VERY LATE reply. I wasn't getting emails or notifications from Github.

I am using PrismLauncher, and I am currently searching for the log(s).

**Comment by PitchBlackNights** [comment]
So, this time, the freeze happened on the main menu, so it's confirmed not limited to the Iris settings screen. The freeze happened within 5-10 seconds of the `18:28:50` mark (at the bottom).

Here is the log:
[https://mclo.gs/Rsk24PO](https://mclo.gs/Rsk24PO)

**Comment by RaphiMC** [comment]
Can you try with only the minimal mods needed for reproducing the problem

**Comment by PitchBlackNights** [comment]
Yes, that will take a bit to do.

**Comment by PitchBlackNights** [comment]
Never mind, it took 3 minutes. The most minimal that still works is with only IF.

Log (freeze happened around the [18:37:20] mark):
[https://mclo.gs/Sq4Z9Ic](https://mclo.gs/Sq4Z9Ic)

**Comment by RaphiMC** [comment]
Can you check if changing launcher (Preferrable not a MultiMC fork, for more accurate results) fixes the issue. I had another report a very similar issue and they also used Prism

**Comment by hexadecimal233** [comment]
Could you make minecraft use `Nvidia GPU`? The log shows minecraft is running on core gpu

(guess this will probably fix the issue)

**Comment by Raconteur32** [comment]
> Could you make minecraft use `Nvidia GPU`? The log shows minecraft is running on core gpu
> 
> (guess this will probably fix the issue)

Ok, since I had the same problem, I tested on my side, and it seems to be the problem. Since I updated my Nvidia drivers, Minecraft with prism have switched to use the integrated GPU. I made it use the Nvidia GPU again and the freeze is gone.

But the question is why IF bug with integrated GPU now.

EDIT: Also, let's wait the answer of the OP, but I think it should work for them to.

**Comment by PitchBlackNights** [comment]
I'm so sorry for not seeing this! My GitHub won't send me notifications, and I've been busy with school. I'll try it, shouldn't take long.

**Comment by PitchBlackNights** [comment]
```
Bug happens with...
=======================================
Integrated Graphics + Prism	TRUE
Graphics Card + Prism		FALSE
Integrated Graphics + Curse	N/A
Graphics Card + Curse		FALSE
Graphics Card? + GDLauncher	FALSE?
```

I couldn't get it to happen with other launchers cause they all forced Minecraft to use my Graphics card. And on GDLauncher, I think it was using my Graphics Card, but there was no log entry to confirm it. All of those tests were made with the default settings that Prism Launcher offers, `-Xmx4096m -Xms512m`, and the only mod installed was IF.

This is a weird edge case. "The Minecraft Render Thread hangs when using IF on 1.19.4 with Fabric 0.14.22 with Intel Integrated UHD Graphics when a Minecraft Sever, running in the background, starts the Java ZGC."

**Comment by RaphiMC** [comment]
Thanks for narrowing it down. I will do my own testing and look for a possible fix

**Comment by RaphiMC** [comment]
Fixed in ImmediatelyFast 1.1.28

**Comment by PitchBlackNights** [comment]
Thanks! Great mod btw.
